### PR TITLE
Validate inbound substreams in `SyncingEngine`

### PR DESCRIPTION
**Pull request by altonen**
Validate inbound substreams before accepting them in `Notifications`. This is done to prevent the constant influx of substreams that are incorrectly being opened by remote nodes because they were first accepted and then rejected.

This is a temporary change but models the upcoming `NotificationService`. Before informing `Peerset` of the new inbound substream, pass it to `SyncingEngine` which first checks whether it can accept the peer. When the response is received from `SyncingEngine`, and if the substream was accepted, inform `Peerset` of the substream which allocates a slot for the connection. When `Peerset` is done with its validation, emit `NotificationStreamOpened` to inform all protocols that a new substream has been opened.

This done to prevent remote peers getting confused about outbound substream first getting accepted and then getting rejected right after.

---
This PR combined with peer eviction yields very promising results. I was running a validator yesterday all day and there was no stalling of any kind. Nevertheless, this is only a temporary change to keep `sc-network` on life support until we've finished the `Peerset`/`Notifications` refactoring.

### CODE REVIEW

File: client/network/sync/src/engine.rs [hunk 1]
```diff
@@ -764,73 +768,90 @@ where
...
+		&mut self,
+		who: PeerId,
+		status: &BlockAnnouncesHandshake<B>,
+		sink: NotificationsSink,
+	) -> Result<(), ()> {
+		log::trace!(target: "sync", "New peer {} {:?}", who, status);
+
+		// peer is not reported in `validate_inbound_substream()` as the peer might be
+		// in `Incoming` state and genesis mismatch would disconnect the peer which is
+		// an invalid state transition in `Notifications`.
```
Comment (dmitry-markin) [inline client/network/sync/src/engine.rs:849]: Could you elaborate on this comment? I'm not sure that I understand it.

File: client/network/src/protocol.rs [hunk 2]
```diff
@@ -413,6 +462,23 @@ impl<B: BlockT> NetworkBehaviour for Protocol<B> {
...
+			match validation_result {
+				Ok((peer, index)) => {
+					self.behaviour.accept_inbound_substream(peer, index);
+				},
+				Err(index) => {
+					log::debug!(
+						target: "sub-libp2p",
+						"`SyncingEngine` rejected stream"
+					);
+					self.behaviour.peerset_report_reject(index);
```
Comment (dmitry-markin) [inline client/network/src/protocol.rs:477]: May be introduce `self.behaviour.reject_inbound_substream(index)`, supplementing `accept_inbound_substream`, to not confuse readers with `peerset` reference here?

**Review by dmitry-markin — APPROVED** [review]

File: client/network/src/protocol/notifications/behaviour.rs [hunk 3]
```diff
@@ -1544,14 +1560,26 @@ impl NetworkBehaviour for Notifications {
...
-								self.peerset.incoming(set_id, peer_id, incoming_id);
+
 								self.incoming.push(IncomingPeer {
 									peer_id,
 									set_id,
 									alive: true,
 									incoming_id,
 								});
 
+								if set_id == 0.into() {
```
Comment (melekes) [inline client/network/src/protocol/notifications/behaviour.rs:1571]: should `0` be a const?

**Review by melekes — APPROVED** [review]
👍

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2679717

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2679893

File: client/network/sync/src/engine.rs [hunk 4]
```diff
@@ -77,8 +77,8 @@ const MAX_KNOWN_BLOCKS: usize = 1024; // ~32kb per peer + LruHashSet overhead
 
 mod rep {
 	use sc_peerset::ReputationChange as Rep;
-	/// Peer has different genesis.
-	pub const GENESIS_MISMATCH: Rep = Rep::new_fatal("Genesis mismatch");
+	// /// Peer has different genesis.
+	// pub const GENESIS_MISMATCH: Rep = Rep::new_fatal("Genesis mismatch");
```
[the remark this answers is not in the document]
Comment (altonen) [reply client/network/sync/src/engine.rs:81]: We need to re-enable this behavior soon so I think it's better to just comment it out for now

**Comment by altonen** [comment]
I think I will close this. CI failed for `sync_after_missing_announcement`. I've run the test all weekend and I can't reproduce it but there is something wrong and considering that the state of syncing is starting to feel hopeless, I'd rather not merge something that could possibly make things worse since I don't know what the underlying issue here is.

**Pull request closed without merging**